sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/EmptyCollectionsShouldNotBeEnumerated.cs

```diff
@@ -0,0 +1,306 @@
+                switch (instruction.Kind())
+                {
+                    case SyntaxKind.ObjectCreationExpression:
+                        var objectCreation = (ObjectCreationExpressionSyntax)instruction;
+                        var constructorSymbol = semanticModel.GetSymbolInfo(objectCreation).Symbol as IMethodSymbol;
+
+                        if (IsCollectionConstructor(constructorSymbol) &&
+                            (!constructorSymbol.Parameters.Any()
+                            || constructorSymbol.Parameters.Count(p => p.IsType(KnownType.System_Int32)) == 1))
+                        {
+                            newProgramState = symbolicValue.SetConstraint(CollectionConstraint.Empty, newProgramState);
+                        }
+                        if (objectCreation.Initializer != null)
+                        {
+                            var constraint = objectCreation.Initializer.Expressions.Count == 0
+                                ? CollectionConstraint.Empty
+                                : CollectionConstraint.NotEmpty;
+                            newProgramState = symbolicValue.SetConstraint(constraint, newProgramState);
+                        }
+                        break;
+                    case SyntaxKind.ArrayCreationExpression:
+                        var arrayCreation = (ArrayCreationExpressionSyntax)instruction;
+                        if (arrayCreation.Type.RankSpecifiers.Count == 1 &&
+                            arrayCreation.Type.RankSpecifiers[0].Sizes.Count == 1)
+                        {
+                            var size = arrayCreation.Type.RankSpecifiers[0].Sizes[0] as LiteralExpressionSyntax;
+                            if (size?.Token.ValueText == "0")
+                            {
+                                newProgramState = symbolicValue.SetConstraint(CollectionConstraint.Empty, newProgramState);
+                            }
+                        }
+                        if (arrayCreation.Initializer != null)
+                        {
+                            var constraint = arrayCreation.Initializer.Expressions.Count == 0
+                                ? CollectionConstraint.Empty
+                                : CollectionConstraint.NotEmpty;
+                            newProgramState = symbolicValue.SetConstraint(constraint, newProgramState);
+                        }
+                        break;
+                    default:
+                        break;
+                }
+                return newProgramState;
+            }
+
+            private static bool IsDictionarySetItem(ElementAccessExpressionSyntax instruction) =>
+                (instruction.GetSelfOrTopParenthesizedExpression().Parent as AssignmentExpressionSyntax)
+                    ?.Left.RemoveParentheses() == instruction;
+
+            private static bool IsCollectionConstructor(IMethodSymbol constructorSymbol) =>

```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Consider using more general type 'Microsoft.CodeAnalysis.ISymbol' instead of 'Microsoft.CodeAnalysis.IMethodSymbol'. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS3242)